src/main/java/com/flipkart/zjsonpatch/JsonPointer.java

```diff
@@ -0,0 +1,313 @@
+ *
+ * <p></p>Generally, a JSON Pointer is a string representation of a path into a JSON document.
+ * This class implements the RFC as closely as possible, and offers several helpers and
+ * utility methods on top of it:
+ *
+ * <pre>
+ *      // Parse, build or render a JSON pointer
+ *      String path = "/a/0/b/1";
+ *      JsonPointer ptr1 = JsonPointer.{@link #parse}(path);
+ *      JsonPointer ptr2 = JsonPointer.{@link #ROOT}.at("a").at(0).at("b").at(1);
+ *      assert(ptr1.equals(ptr2));
+ *      assert(path.equals(ptr1.toString()));
+ *      assert(path.equals(ptr2.toString()));
+ *
+ *      // Evaluate a JSON pointer against a live document
+ *      ObjectMapper om = new ObjectMapper();
+ *      JsonNode doc = om.readTree("{\"foo\":[\"bar\", \"baz\"]}");
+ *      JsonNode baz = JsonPointer.parse("/foo/1").{@link #evaluate(JsonNode) evaluate}(doc);
+ *      assert(baz.textValue().equals("baz"));
+ * </pre>
+ *
+ * <p>Instances of {@link JsonPointer} and its constituent {@link RefToken}s are <b>immutable</b>.
+ *
+ * @since 0.4.8
+ */
+class JsonPointer {
+    private final RefToken[] tokens;
+
+    /** A JSON pointer representing the root node of a JSON document */
+    public final static JsonPointer ROOT = new JsonPointer(new RefToken[] {});
+
+    private JsonPointer(RefToken[] tokens) {
+        this.tokens = tokens;
+    }
+
+    /**
+     * Constructs a new pointer from a list of reference tokens.
+     *
+     * @param tokens The list of reference tokens from which to construct the new pointer. This list is not modified.
+     */
+    public JsonPointer(List<RefToken> tokens) {
+        this.tokens = tokens.toArray(new RefToken[0]);
+    }
+
+    private static Pattern JSON_POINTER_PATTERN = Pattern.compile("\\G/(.*?)(?=/|\\z)");
+
+    /**
+     * Parses a valid string representation of a JSON pointer.
+     *
+     * @param path The string representation to be parsed.
+     * @return An instance of {@link JsonPointer} conforming to the specified string representation.
+     */
+    public static JsonPointer parse(String path) {
+        Matcher matcher = JSON_POINTER_PATTERN.matcher(path);
+        List<RefToken> result = new ArrayList<RefToken>();
+        while (matcher.find()) {
+            result.add(RefToken.parse(matcher.group(1)));
+        }
+        if (result.isEmpty()) return ROOT;
+        return new JsonPointer(result);
+    }
+
+    /**
+     * Indicates whether or not this instance points to the root of a JSON document.
+     * @return {@code true} if this pointer represents the root node, {@code false} otherwise.
+     */
+    public boolean isRoot() {
+        return tokens.length == 0;
+    }
+
+    /**
+     * Creates a new JSON pointer to the specified field of the object referenced by this instance.
+     *
+     * @param field The desired field name, or any valid JSON Pointer reference token
+     * @return The new {@link JsonPointer} instance.
+     */
+    JsonPointer at(String field) {
```

Comment:
Nitpick: Maybe "append" instead of "at" ? more verbose but more semantically correct

Reply:
No objection, will change.